# Data Cleaning

## Objective
The objective of this notebook is to prepare the dataset for the exploratory analysis.

The cleaning process includes:
- Removing observations with missing values in essential metadata.
- Removing duplicated songs based on `track_id`.
- Verifying the consistency of the cleaned dataset.
- Saving the cleaned dataset for subsequent analysis.

In [27]:
import pandas as pd
path = "../data/raw/spotify_songs.csv"

df_original = pd.read_csv(path)

### Removing missing values

In [28]:
rows_before=len(df_original)

df = df_original.dropna()
rows_after=len(df)
print(f"Rows before: {rows_before}\nRows after: {rows_after}\nRows dropped: {rows_before-rows_after}")

Rows before: 32833
Rows after: 32828
Rows dropped: 5


### Remove duplicated songs
The dataset contains multiple occurrences of the same song because a track can belong to multiple playlists.
Since the objective of this project is to analyse songs rather than playlists, duplicated `track_id` values are removed, while keeping the first occurrence.


In [29]:
rows_before=len(df)
df = df.drop_duplicates(subset ='track_id', keep = 'first')
rows_after=len(df)

print(f"Rows before: {rows_before}\nRows after: {rows_after}\nDuplicates removed: {rows_before-rows_after}")


Rows before: 32828
Rows after: 28352
Duplicates removed: 4476


In [33]:
missing_track_ids = df_original[df_original.isna().any(axis=1) & df_original.duplicated(subset='track_id', keep='first')]["track_id"]
print(missing_track_ids)

19811    69gRFGOWY9OMpFJgFol1u0
Name: track_id, dtype: str


In [ ]:
df_original[df_original.track_id == '69gRFGOWY9OMpFJgFol1u0']

,track_id,track_name,track_artist,track_popularity,track_album_id,track_album_name,track_album_release_date,playlist_name,playlist_id,playlist_genre,...,key,loudness,mode,speechiness,acousticness,instrumentalness,liveness,valence,tempo,duration_ms
8151,69gRFGOWY9OMpFJgFol1u0,NaN,NaN,0,717UG2du6utFe7CdmpuUe3,NaN,2012-01-05,HIP&HOP,5DyJsJZOpMJh34WvUrQzMV,rap,...,6,-7.635,1,0.176,0.041,0.0,0.116,0.649,95.999,282707
19811,69gRFGOWY9OMpFJgFol1u0,NaN,NaN,0,717UG2du6utFe7CdmpuUe3,NaN,2012-01-05,latin hip hop,3nH8aytdqNeRbcRCg3dw9q,latin,...,6,-7.635,1,0.176,0.041,0.0,0.116,0.649,95.999,282707


### Save cleaned dataset

In [38]:
output_path = "../data/processed/spotify_songs_cleaned.csv"
df.to_csv(output_path, index=False)

In [40]:
pd.read_csv(output_path).head()

,track_id,track_name,track_artist,track_popularity,track_album_id,track_album_name,track_album_release_date,playlist_name,playlist_id,playlist_genre,...,key,loudness,mode,speechiness,acousticness,instrumentalness,liveness,valence,tempo,duration_ms
0,6f807x0ima9a1j3VPbc7VN,I Don't Care (with Justin Bieber) - Loud Luxur...,Ed Sheeran,66,2oCs0DGTsRO98Gh5ZSl2Cx,I Don't Care (with Justin Bieber) [Loud Luxury...,2019-06-14,Pop Remix,37i9dQZF1DXcZDD7cfEKhW,pop,...,6,-2.634,1,0.0583,0.1020,0.000000,0.0653,0.518,122.036,194754
1,0r7CVbZTWZgbTCYdfa2P31,Memories - Dillon Francis Remix,Maroon 5,67,63rPSO264uRjW1X5E6cWv6,Memories (Dillon Francis Remix),2019-12-13,Pop Remix,37i9dQZF1DXcZDD7cfEKhW,pop,...,11,-4.969,1,0.0373,0.0724,0.004210,0.3570,0.693,99.972,162600
2,1z1Hg7Vb0AhHDiEmnDE79l,All the Time - Don Diablo Remix,Zara Larsson,70,1HoSmj2eLcsrR0vE9gThr4,All the Time (Don Diablo Remix),2019-07-05,Pop Remix,37i9dQZF1DXcZDD7cfEKhW,pop,...,1,-3.432,0,0.0742,0.0794,0.000023,0.1100,0.613,124.008,176616
3,75FpbthrwQmzHlBJLuGdC7,Call You Mine - Keanu Silva Remix,The Chainsmokers,60,1nqYsOef1yKKuGOVchbsk6,Call You Mine - The Remixes,2019-07-19,Pop Remix,37i9dQZF1DXcZDD7cfEKhW,pop,...,7,-3.778,1,0.1020,0.0287,0.000009,0.2040,0.277,121.956,169093
4,1e8PAfcKUYoKkxPhrHqw4x,Someone You Loved - Future Humans Remix,Lewis Capaldi,69,7m7vv9wlQ4i0LFuJiE2zsQ,Someone You Loved (Future Humans Remix),2019-03-05,Pop Remix,37i9dQZF1DXcZDD7cfEKhW,pop,...,1,-4.672,1,0.0359,0.0803,0.000000,0.0833,0.725,123.976,189052


## Summary
The following cleaning steps were performed:
- Removed **5 observations** containing missing values.
- Removed **4476 duplicated song occurrences** based on `track_id`, keeping the first occurrence of each song. 
-Verified that the observation with the `track_id 69gRFGOWY9OMpFJgFol1u0` was duplicated and had missing values in both of its occurrences. 
- Exported the cleaned dataset to `data/processed/spotify_songs_clean.csv`.

The cleaned dataset contains **28352 unique songs** and is ready for exploratory analysis.
